### PR TITLE
webtau-726 compute success and error branch coverage

### LINKED ISSUE #726: Include success and error outcomes in GraphQL Coverage
In contrast to OpenAPI where we can declare possible outcomes in the schema itself (e.g. 200, 201, 302, 400, 412, default), graphQL responses have either a “happy” outcome with the `data` property populated (and `errors` being null) XOR an error outcome where `data` is null and `errors` is populated.

We can enhance graphql support in webtau by saying the set of “declared queries” consists of both the happy and error outcomes of all queries declared in the schema.

Recording results would have to inspect the response body and determine happiness based on the data/errors properties.

100% test coverage then means that every declared query in our schema has been hit with at least 1 successful and 1 error outcome.

Alternative:
In addition to `graphQLSkippedQueries` and `graphQLCoveredQueries`, we can introduce `ReportCustomData` for `graphQLCoveredQueriesWithData` and `graphQLCoveredQueriesWithErrors`. That would give us the opportunity to introduce something similar to "branch coverage" vs "method coverage" where we can add `covered happy paths = graphQLCoveredQueriesWithData / declaredQueries` and `covered error paths = graphQLCoveredQueriesWithErrors / declaredQueries`

### PR DESCRIPTION
**Pull request by tsiq-clemens**
Fixes #726 
Inspect the validation result to determine whether it was a success (no errors) or error outcome (errors != null), then use that information to gather and calculate success/error "branch" coverage for all defined graphql operations.

### CODE REVIEW

File: webtau-feature-testing/pom.xml [hunk 1]
```diff
@@ -136,6 +136,7 @@
                 <artifactId>spring-boot-maven-plugin</artifactId>
                 <configuration>
                     <mainClass>com.example.demo.springboot.app.SpringBootDemoApp</mainClass>
+                    <wait>40000</wait>
```
Comment (tsiq-clemens) [inline webtau-feature-testing/pom.xml:139]: The default of 30s was not enough in my local development. Happy to revert this change if it works for everyone else.

**Comment by tsiq-clemens** [comment]
One of the main challenges here is naming -- what say you about calling this "branch coverage"?

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java [hunk 2]
```diff
@@ -38,16 +38,34 @@ public void recordQuery(HttpValidationResult validationResult) {
...
         Set<GraphQLQuery> graphQLQueries = schema.findQueries(validationResult.getRequestBody());
-        graphQLQueries.forEach(query -> coveredQueries.add(query, validationResult.getId(), validationResult.getElapsedTime()));
+        graphQLQueries.forEach(query -> coveredQueries.
+            add(query, validationResult.getId(), validationResult.getElapsedTime(), isErrorResult(validationResult)));
     }
 
-    Stream<GraphQLQuery> nonCoveredQueries() {
-        return schema.getSchemaDeclaredQueries().filter(o -> !coveredQueries.contains(o));
+    private boolean isErrorResult(HttpValidationResult validationResult) {
+        return JsonUtils.convertToTree(JsonUtils.deserialize(validationResult.getResponse().getTextContent()))
```
Comment (tsiq-karold) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:46]: Could this throw? If it can, we probably should catch, log and move on without throwing.

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java [hunk 3]
```diff
@@ -38,16 +38,34 @@ public void recordQuery(HttpValidationResult validationResult) {
...
+      }
+
     Stream<GraphQLQuery> coveredQueries() {
         return coveredQueries.coveredQueries();
     }
+    Stream<GraphQLQuery> coveredSuccessBranches() {
+        return coveredQueries.coveredSuccessBranches();
+    }
+    Stream<GraphQLQuery> nonCoveredSuccessBranches() {
+        return schema.getSchemaDeclaredQueries().filter(o -> coveredQueries.coveredSuccessBranches().noneMatch(graphQLQuery -> graphQLQuery.equals(o)));
```
Comment (tsiq-karold) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:61]: I suggest sticking `coveredQueries.coveredSuccessBranches()` into a local var inside this function (not in the lambda) otherwise you're doing the computation/filtration from `coveredSuccessBranches()` repeatedly.

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java [hunk 4]
```diff
@@ -38,16 +38,34 @@ public void recordQuery(HttpValidationResult validationResult) {
...
+    }
+    Stream<GraphQLQuery> coveredErrorBranches() {
+        return coveredQueries.coveredErrorBranches();
+    }
+    Stream<GraphQLQuery> nonCoveredErrorBranches() {
+        return schema.getSchemaDeclaredQueries().filter(o -> coveredQueries.coveredErrorBranches().noneMatch(graphQLQuery -> graphQLQuery.equals(o)));
```
Comment (tsiq-karold) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:67]: Same comment as on nonCoveredSuccessBranches.

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLReportDataProvider.java [hunk 5]
```diff
@@ -53,15 +53,35 @@ public GraphQLReportDataProvider(Supplier<GraphQLCoverage> coverageSupplier) {
                 .map(GraphQLQuery::toMap)
                 .collect(Collectors.toList());
 
+        List<? extends Map<String, ?>> successBranches = coverageSupplier.get().coveredSuccessBranches()
```
Comment (tsiq-karold) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLReportDataProvider.java:56]: Nothing you introduced here and not to be addressed in this PR, more a note to ourselves, but we should really find a better way of handling the stuff in this function.  It's sooo repetitive.

**Comment by tsiq-karold** [comment]
> One of the main challenges here is naming -- what say you about calling this "branch coverage"?

I'm happy with the naming.

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLReportDataProvider.java (hunk 5 above)
Comment (tsiq-clemens) [reply webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLReportDataProvider.java:56]: we can at least do a little better, agreed

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java (hunk 2 above)
Comment (tsiq-clemens) [reply webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:47]: Very interesting point.
There seems to be no standard logging in reporters or elsewhere. I agree that we should be able to move on if something goes wrong here. I'd say if an error cannot be detected, it's not an `errorResult`, but how visible is a stacktrace or other error log here?

Comment (tsiq-karold) [reply webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:49]: There is some precedent for ignoring these sorts of issues: https://github.com/testingisdocumenting/webtau/blob/master/webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLSchema.java#L68 but you're right, it's generally silent which is not ideal.

**Review by tsiq-karold — APPROVED** [review]

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java [hunk 6]
```diff
@@ -38,16 +40,41 @@ public void recordQuery(HttpValidationResult validationResult) {
...
+        return JsonUtils.convertToTree(JsonUtils.deserialize(validationResult.getResponse().getTextContent()))
+            .has("errors");
+      } catch (Exception e) {
+        e.printStackTrace();
+        return false;
+      }
     }
 
+    Stream<GraphQLQuery> nonCoveredQueries() {
+          return schema.getSchemaDeclaredQueries().filter(o -> !coveredQueries.contains(o));
```
Comment (MykolaGolubyev) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:58]: src text misalignment

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java [hunk 7]
```diff
@@ -38,16 +40,41 @@ public void recordQuery(HttpValidationResult validationResult) {
...
     }
 
-    Stream<GraphQLQuery> nonCoveredQueries() {
-        return schema.getSchemaDeclaredQueries().filter(o -> !coveredQueries.contains(o));
+    private boolean isErrorResult(HttpValidationResult validationResult) {
+      try {
+        return JsonUtils.convertToTree(JsonUtils.deserialize(validationResult.getResponse().getTextContent()))
+            .has("errors");
+      } catch (Exception e) {
+        e.printStackTrace();
```
Comment (MykolaGolubyev) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:52]: do we need to see this printed stack trace?

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java [hunk 8]
```diff
@@ -38,16 +40,41 @@ public void recordQuery(HttpValidationResult validationResult) {
...
     }
 
+    Stream<GraphQLQuery> nonCoveredQueries() {
+          return schema.getSchemaDeclaredQueries().filter(o -> !coveredQueries.contains(o));
+      }
+
     Stream<GraphQLQuery> coveredQueries() {
         return coveredQueries.coveredQueries();
     }
+    Stream<GraphQLQuery> coveredSuccessBranches() {
```
Comment (MykolaGolubyev) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:64]: we don't have automated formatting test, sorry for nitpicking,
could you please separate methods with an empty line?

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLReportDataProvider.java [hunk 9]
```diff
@@ -45,23 +45,28 @@ public GraphQLReportDataProvider(Supplier<GraphQLCoverage> coverageSupplier) {
...
         List<? extends Map<String, ?>> timingByQuery = computeTiming();
-
         Map<String, ?> coverageSummary = computeCoverageSummary();
 
-        return Stream.of(
-                new ReportCustomData("graphQLSkippedQueries", nonCoveredQueries),
-                new ReportCustomData("graphQLCoveredQueries", coveredQueries),
-                new ReportCustomData("graphQLQueryTimeStatistics", timingByQuery),
-                new ReportCustomData("graphQLCoverageSummary", coverageSummary));
+    return Stream.of(
```
Comment (MykolaGolubyev) [inline webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLReportDataProvider.java:57]: formatting misalign

**Review by MykolaGolubyev — COMMENTED** [review]
Like it! thank you.
when you are ready to merge, make sure commit message is prefixed with 
`graphql: <your message>`

**Comment by tsiq-clemens** [comment]
fixed the formatting.

File: webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java (hunk 7 above)
Comment (tsiq-clemens) [reply webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/GraphQLCoverage.java:52]: I'd expect this to be a very rare occurrence, and then the detail might be useful.
